### PR TITLE
Issue 22 consensus fairness

### PR DESCRIPTION
**Pull request by SpencerDawkins**

### CODE REVIEW

**Comment by SpencerDawkins** [comment]
@LPardue - is this roughly what you were expecting to see, on Issue 22? 

Please advise :-)

File: draft-ietf-mops-streaming-opcons.md [hunk 1]
```diff
@@ -219,6 +288,46 @@ Presentations:
...
+
+In recent times, the usage of UDP-based applications that were not simple query-response protocols has grown substantially, and since UDP does not provide any feedback mechanism to senders to help limit impacts on other users, application-level protocols such as RTP {{RFC3550}} have been responsible for the decisions that TCP-based applications have delegated to TCP - what to send, how much to send, and when to send it. So, the way these UDP-based applications interact with other users has changed.
+
+It's also worth pointing out that because UDP has no transport-layer feedback mechanisms, UDP-based applications that expect to send and receive substantial amounts of information must provide their own feedback mechanisms. RTP replies on RTCP Sender and Receiver Reports {{RFC3550}} as its own feedback mechanism, and even includes Circuit Breakers for Unicast RTP Sessions {{RFC8083}} for situations when normal RTP congestion control has not been able to react sufficiently to RTP flows sending at rates that result in sustained packet loss.
```
Comment (LPardue) [inline draft-ietf-mops-streaming-opcons.md:303]:
```suggestion
It's also worth pointing out that because UDP has no transport-layer feedback mechanisms, UDP-based applications that expect to send and receive substantial amounts of information must provide their own feedback mechanisms. RTP relies on RTCP Sender and Receiver Reports {{RFC3550}} as its own feedback mechanism, and even includes Circuit Breakers for Unicast RTP Sessions {{RFC8083}} for situations when normal RTP congestion control has not been able to react sufficiently to RTP flows sending at rates that result in sustained packet loss. 
```

maybe a typo?

File: draft-ietf-mops-streaming-opcons.md [hunk 2]
```diff
@@ -219,6 +288,46 @@ Presentations:
  * MOPS 2020-10-30 Interim meeting:\\
    https://www.youtube.com/watch?v=vDZKspv4LXw&t=17m15s
 
+#Evolution of Transport Protocols and Transport Protocol Behaviors
+
+Because networking resources are shared between users, a good place to start our discussion is how contention between users, and mechanisms to resolve that contention in ways that are "fair" between users, impact streaming media users. These topics are closely tied to the usage of transport protocols and transport protocol behaviors. 
+
+For most of the history of the Internet, the dominant transport protocols in use have been UDP and TCP, and they have each had relatively consistent behaviors, although those behaviors have changed over time. 
+
+## UDP and UDP Protocol Behaviors {#udp-behavior}
```
Comment (LPardue) [inline draft-ietf-mops-streaming-opcons.md:297]: nit: there's a bit of redundancy here and elsewhere in the PR with e.g. "UDP Protocol"

File: draft-ietf-mops-streaming-opcons.md [hunk 3]
```diff
@@ -219,6 +288,46 @@ Presentations:
...
+
+The notion of "Circuit Breakers" has also been applied to other UDP applications, such as tunneling packets that are potentially not congestion-controlled over UDP, in {{RFC8084}}.
+
+## TCP and TCP Protocol Behaviors {#tcp-behavior}
+
+For most of the history of the Internet, we have trusted the TCP protocol to limit the impact of applications that sent a significant number of packets, in either or both directions, on other users. Although early versions of TCP were not particularly good at limiting this impact {{RFC0793}}, the addition of Slow Start and Congestion Avoidance, as described in {{RFC2001}}, were critical in allowing TCP-based applications to "use as much bandwidth as possible, but to avoid using more bandwidth than was possible". Although dozens of RFCs have been written refining TCP decisions about what to send, how much to send, and how when to send it, since 1988 {{Jacobson-Karels}} the signals available for TCP senders remained unchanged - end-to-end acknowledgments for packets that were successfully sent and received, and packet timeouts for packets that were not.
```
Comment (LPardue) [inline draft-ietf-mops-streaming-opcons.md:309]:
```suggestion
For most of the history of the Internet, we have trusted the TCP protocol to limit the impact of applications that sent a significant number of packets, in either or both directions, on other users. Although early versions of TCP were not particularly good at limiting this impact {{RFC0793}}, the addition of Slow Start and Congestion Avoidance, as described in {{RFC2001}}, were critical in allowing TCP-based applications to "use as much bandwidth as possible, but to avoid using more bandwidth than was possible". Although dozens of RFCs have been written refining TCP decisions about what to send, how much to send, and when to send it, since 1988 {{Jacobson-Karels}} the signals available for TCP senders remained unchanged - end-to-end acknowledgments for packets that were successfully sent and received, and packet timeouts for packets that were not. 
```

File: draft-ietf-mops-streaming-opcons.md [hunk 4]
```diff
@@ -219,6 +288,46 @@ Presentations:
...
+
+The success of the largely TCP-based Internet is evidence that the mechanisms used to achieve equilibrium quickly, at a point where TCP senders do not interfere with other TCP senders for sustained periods of time, have been largely successful, even if the specific mechanisms used to reach equilibrium have change over time. Because TCP provides a common tool to avoid contention, new TCP-based applications (for example, HTTP) have had the same transport behavior as older TCP-based applications (for example, FTP).
```
Comment (LPardue) [inline draft-ietf-mops-streaming-opcons.md:311]: the sentences here are a bit long. Would recommend to break them up a little.

I also personally wouldn't call HTTP a new protocol :D

File: draft-ietf-mops-streaming-opcons.md [hunk 5]
```diff
@@ -219,6 +288,46 @@ Presentations:
...
+
+## The QUIC Protocol and QUIC Protocol Behavior
+
+The QUIC protocol {{I-D.ietf-quic-transport}}, developed from a proprietary protocol into an IETF standards-track protocol {{I-D.ietf-quic-transport}}, turns many of the statements made in {{udp-behavior}} and {{tcp-behavior}} on their heads. 
+
+The standardized QUIC protocol is used to carry HTTP traffic, but instead of being TCP-based, HTTP/3 {{I-D.ietf-quic-http}} is encapsulated in QUIC, which is then encapsulated in UDP, so streaming operators (and network operators) will see traffic that looks like HTTP, but is carried over UDP (specifically, using UDP port 443 {{Port443}}). Network operators may already be blocking UDP traffic on this port, since HTTP over TLS/SSL has used TCP, not UDP. Even if UDP traffic using this port is not blocked, traffic using this port may be severely rate-limited, since HTTP/3 over QUIC has the potential to send much more traffic over UDP than the network operator expects.
```
Comment (LPardue) [inline draft-ietf-mops-streaming-opcons.md:321]: As someone overly familiar with HTTP/3 this paragraph reads a bit strangely. 

I might rephrase it to something like below (and defer to your editor capabilites to make it even better)

The QUIC transport protocol can carry different application-later protocols. HTTP/3 {{I-D.ietf-quic-http}} describes how QUIC transport features are used for HTTP. The convention is for HTTP/3 to run over UDP port 443 {{Port443}} but this is not a strict requirement. Since earlier versions of HTTP(S) rely on TCP, UDP ports may blocked for any values that are not commonly used, such as UDP 53 for DNS. Where UDP ports are not blocked and HTTP/3 can flow, streaming operators (and network operators) might see QUIC traffic patterns that are similar to HTTP(S) over TCP. Where ports are open, they may be servery rate-limited because operators may not be used to seeing legitimate high-bandwidth traffic such as streaming, over the ports that HTTP/3 tends to use.

File: draft-ietf-mops-streaming-opcons.md [hunk 6]
```diff
@@ -219,6 +288,46 @@ Presentations:
...
+
+As noted elsewhere in this document, the QUIC protocol encrypts almost all of its transport parameters, and all of its payload, so any intermediaries that network operators may be using to perform analytics or even participate in current HTTP conversations will not work for HTTP/3 without making changes to their networks.
+
+As noted in {{tcp-behavior}}, there is increasing interest in transport protocol behaviors that responds to delay measurements, instead of responding to packet loss. These behaviors may deliver improved user experience, but in some cases have not responded to sustained packet loss, which exhausts available buffers along the end-to-end path that may affect other users sharing that path. The standardized QUIC protocol includes a congestion-control mechanism {{I-D.ietf-quic-recovery}} that is loss-based, and is intended to have roughly the behavior of TCP NewReno {{RFC6582}}, but the signals QUIC provides for congestion control are intended to be generic, and a sender can unilaterally chose a different algorithm to use, which might be loss-based, like CUBIC {{RFC8312}}, a delay-based congestion controller like COPA or BBR, or something completely different.
```
Comment (LPardue) [inline draft-ietf-mops-streaming-opcons.md:325]: I would invert the sentences about QUIC a a little and say:

QUIC {{I-D.ietf-quic-transport}} provides a set of congestion control hooks that can be use for algorithm agility. {{I-D.ietf-quic-recovery}} defines a basic algorithm that behaves roughly similar TCP NewReno {{RFC6582}}. However, QUIC senders can and do unilaterally chose to use different algorithms such as loss-based CUBIC {{RFC8312}}, delay-based COPA or BBR, or even something completely different

File: draft-ietf-mops-streaming-opcons.md (hunk 5 above)
Comment (LPardue) [reply draft-ietf-mops-streaming-opcons.md:321]: how about adding some links to the QUIC operability draft?

File: draft-ietf-mops-streaming-opcons.md [hunk 7]
```diff
@@ -219,6 +288,46 @@ Presentations:
...
+
+The decision to deploy a new version of QUIC is relatively uncontrolled, compared to other widely used transport protocols, and "a new version of QUIC" can include "a new congestion controller", so QUIC protocol behavior can change quickly, without much notice. At IETF 105, Christian Huitema and Brian Trammell presented a talk on "Congestion Defense in Depth" {{CDiD}}, that explored potential concerns about new QUIC congestion controllers being broadly deployed without the testing and instrumentation that current major content providers routinely include. The sense of the room at IETF 105 was that the current major content providers understood what is at stake when they deploy new congestion controllers, but this presentation, and the related discussion in TSVAREA minutes from IETF 105 ({{tsvarea-105}}, are still worth a look from new and rapidly growing content providers.
```
Comment (LPardue) [inline draft-ietf-mops-streaming-opcons.md:329]: I think there is a sublte point that is hard to make here. I'd be surprised if operators rolled a QUIC version just to deploy a different CC - that is the beauty of sender unilateral descisions. However, what we might see is that protocol evolution can support new or better approaches to CC. For example, Christians's proposed one-way delay measurement extension. Tangentially, some experimenters might want to add transport feautres that need alternative CC, and it can be easier to run an experimental version of QUIC to prove a hypothesis rather than try to hook into extension mechanisms.

I really don't know best to present such nuance here. All I can suggest is to focus on concerns your target audience has, rather than nitty gritty QUIC detail.s

File: draft-ietf-mops-streaming-opcons.md (hunk 2 above)
Comment (SpencerDawkins) [reply draft-ietf-mops-streaming-opcons.md:297]: Yeah, I was trying to say something that didn't matter. This will be "UDP and Its behavior", and elsewhere.

File: draft-ietf-mops-streaming-opcons.md (hunk 4 above)
Comment (SpencerDawkins) [reply draft-ietf-mops-streaming-opcons.md:311]: Yeah, I can do that. Coincidentally, I downloaded the diction package yesterday because I was wondering about this. :-)

I can remember when HTTP was a new protocol, but I take your point. The target audience includes people who haven't retired yet :D I'll rephrase this.

File: draft-ietf-mops-streaming-opcons.md [hunk 8]
```diff
@@ -219,6 +288,46 @@ Presentations:
...
+
+In recent times, the TCP goal of probing for available bandwidth, and "backing off" when a network path is saturated, has been supplanted by the goal of avoiding growing queues along network paths, which prevent TCP senders from reacting quickly when a network path is saturated. Congestion control mechanisms such as COPA {{COPA}} and BBR {{I-D.cardwell-iccrg-bbr-congestion-control}} make these decisions based on measured path delays, assuming that if the measured path delay is increasing, the sender is injecting packets onto the network path faster than the receiver can accept them, so the sender should adjust its sending rate accordingly. 
+
+Although TCP protocol behavior has changed over time, the common practice of implementing TCP as part of an operating system kernel has acted to limit how quickly TCP behavior can change, even with the widespread use of automated operating system update installation on many end-user systems, so that streaming media providers could have a reasonable expectation that they could understand TCP transport protocol behaviors, and that those behaviors would remain relatively stable in the short term.
```
[the remark this answers is not in the document]
Comment (SpencerDawkins) [reply draft-ietf-mops-streaming-opcons.md:315]: Not any more :-)

File: draft-ietf-mops-streaming-opcons.md [hunk 9]
```diff
@@ -219,6 +288,46 @@ Presentations:
...
+
+Although TCP protocol behavior has changed over time, the common practice of implementing TCP as part of an operating system kernel has acted to limit how quickly TCP behavior can change, even with the widespread use of automated operating system update installation on many end-user systems, so that streaming media providers could have a reasonable expectation that they could understand TCP transport protocol behaviors, and that those behaviors would remain relatively stable in the short term. 
+
+## The QUIC Protocol and QUIC Protocol Behavior
+
+The QUIC protocol {{I-D.ietf-quic-transport}}, developed from a proprietary protocol into an IETF standards-track protocol {{I-D.ietf-quic-transport}}, turns many of the statements made in {{udp-behavior}} and {{tcp-behavior}} on their heads.
```
[the remark this answers is not in the document]
Comment (SpencerDawkins) [reply draft-ietf-mops-streaming-opcons.md:319]: Thanks for spotting the double reference. 

I'm using these suggestions.

File: draft-ietf-mops-streaming-opcons.md (hunk 2 above)
Comment (LPardue) [reply draft-ietf-mops-streaming-opcons.md:297]: Wfm

File: draft-ietf-mops-streaming-opcons.md (hunk 6 above)
Comment (SpencerDawkins) [reply draft-ietf-mops-streaming-opcons.md:325]: Thanks for this suggestion. I used very close to your text (with minor dorking).

File: draft-ietf-mops-streaming-opcons.md (hunk 5 above)
Comment (SpencerDawkins) [reply draft-ietf-mops-streaming-opcons.md:321]: I started with your text, although I jumped up and down on it a bit. And re: pointing to the manageability draft, duh, I am A Idiot for not thinking of that - thank you for the wake-up call!

**Comment by SpencerDawkins** [comment]
I'm having a conversation with @martinthompson about https://github.com/quicwg/ops-drafts/issues/339, but the short summary here is that I need to tighten up some of the language about what we are saying about strangely HTTPish UDP traffic, from an observer's perspective, and to do so without starting to say things that either duplicate (bad) or contradict (worse) the QUIC manageability draft on this point. You'll note that the QUIC issue is still under discussion, so probably need to revisit this after they close their Issue. 

Fortunately, I bet @LPardue will make sure I behave ... and I appreciate that.

File: draft-ietf-mops-streaming-opcons.md [hunk 10]
```diff
@@ -219,6 +289,52 @@ Presentations:
...
+
+## The QUIC Protocol and Its Behavior
+
+The QUIC protocol, developed from a proprietary protocol into an IETF standards-track protocol {{I-D.ietf-quic-transport}}, turns many of the statements made in {{udp-behavior}} and {{tcp-behavior}} on their heads. 
+
+Although QUIC provides an alternative to the TCP and UDP transport protocols, QUIC is itself encapsulated in UDP. As noted elsewhere in this document, the QUIC protocol encrypts almost all of its transport parameters, and all of its payload, so any intermediaries that network operators may be using to perform analytics or even intercept exchanges in current applications will not work for QUIC-based applications without making changes to their networks.
+
+While QUIC is designed as a general-purpose transport protocol, and can carry different application-later protocols, the current standardized mapping is for HTTP/3 {{I-D.ietf-quic-http}}, which describes how QUIC transport features are used for HTTP. The convention is for HTTP/3 to run over UDP port 443 {{Port443}} but this is not a strict requirement.
```
Comment (LPardue) [inline draft-ietf-mops-streaming-opcons.md:324]:
```suggestion
While QUIC is designed as a general-purpose transport protocol, and can carry different application-layer protocols, the current standardized mapping is for HTTP/3 {{I-D.ietf-quic-http}}, which describes how QUIC transport features are used for HTTP. The convention is for HTTP/3 to run over UDP port 443 {{Port443}} but this is not a strict requirement. 
```

File: draft-ietf-mops-streaming-opcons.md [hunk 11]
```diff
@@ -219,6 +289,52 @@ Presentations:
...
+
+While QUIC is designed as a general-purpose transport protocol, and can carry different application-later protocols, the current standardized mapping is for HTTP/3 {{I-D.ietf-quic-http}}, which describes how QUIC transport features are used for HTTP. The convention is for HTTP/3 to run over UDP port 443 {{Port443}} but this is not a strict requirement. 
+
+When HTTP/3 is encapsulated in QUIC, which is then encapsulated in UDP, this means that streaming operators (and network operators) might see UDP traffic patterns that are similar to HTTP(S) over TCP. Since earlier versions of HTTP(S) rely on TCP, UDP ports may blocked for any port numbers that are not commonly used, such as UDP 53 for DNS. Even when UDP ports are not blocked and HTTP/3 can flow, streaming operators (and network operators) may severely rate-limit this traffic because they do not expect to see legitimate high-bandwidth traffic such as streaming, over the UDP ports that HTTP/3 tends to use.
```
[the remark this answers is not in the document]
Comment (SpencerDawkins) [reply draft-ietf-mops-streaming-opcons.md:326]: @LPardue, thank you for spotting typos!

**Comment by SpencerDawkins** [comment]
> Is it worth also saying about the expectations networks allows streaming traffic over TCP and QUIC to flow freely? I think this could just be a short paragraph that highlights that when other variables are isolated, fairness/contention between TCP and QUIC is most likely a function of the congestion controllers at play.

@LPardue - thank you for this suggestion. See what you think of the current text.

File: draft-ietf-mops-streaming-opcons.md [hunk 12]
```diff
@@ -219,6 +290,54 @@ Presentations:
...
+
+For most of the history of the Internet, the dominant transport protocols in use have been UDP and TCP, and they have each had relatively consistent behaviors, although those behaviors have changed over time. 
+
+## UDP and Its Behavior {#udp-behavior}
+
+For most of the history of the Internet, we have trusted UDP-based applications to limit their impact on other users. One of the strategies used was to use UDP for simple query-response application protocols, such as DNS, which is often used to send a single-packet request to look up the IP address for a DNS name, and return a single-packet response containing the IP address. Although it is possible to saturate a path between a DNS client and DNS server with DNS requests, in practice, that was rare enough that DNS included few mechanisms to resolve contention between DNS users and other users (whether they are also using DNS, or using other application protocols.
```
[the remark this answers is not in the document]
Comment (SpencerDawkins) [reply draft-ietf-mops-streaming-opcons.md:301]: Hi, @GrumpyOldTroll - more broadly, up through at least -02, and maybe -03, it felt like we were describing where streaming media WAS, and now we're adding material that's describing where streaming media is GOING. Does that make sense? 

We didn't punt the issues out of GitHub, and I sure don't disagree with including material on heavily and agile QUIC, which is now approved for publication and being broadly deployed, so saying "come back in a couple of years, and we'll explain what you should have done" is wrong. But it is ... different. 

Re: location - we'll do the right thing. I'm the kind of transport guy that thinks congestion control, contention, and fairness are fundamental, but I'd rather put them in the right place than keep them up front for religious reasons!

**Comment by SpencerDawkins** [comment]
@GrumpyOldTroll I'm going to do a commit that moves this text to an actual reasonable place in the draft (I'll be looking at your suggestion to put it near ABR, and I'll try to Do The Right Thing). 

Probably best not to merge this PR until I commit that, OK?

**Comment by SpencerDawkins** [comment]
@GrumpyOldTroll and @acbegen, I think this PR is ready to merge. Do either of you agree?
